### DISCUSSION TITLE
Docker build for arm64 trips over gflags

### DISCUSSION THREAD

**Discussion by KtGunn** [ifm/ifm3d#387] [Q&A] [answered]
I am building a simple docker image for linux/arm64 on a linux/amd64 host. I am using the latest image from ifm:
ghcr.io.ifm.ifm3d.stable  tag 0.1   ID 33d28a1c458b   size  444MB.
I have a small app that collects images and sends the out over the network. As the image builds it trips over gflags:

![gflags](https://user-images.githubusercontent.com/6865198/222477827-e0e7c063-a409-4ad1-bae1-1c1d1859082c.png)

I have tried adding (sudo) apt-get install libgflags-dev two different ways (see Dockerfile). One way fails to find the package the other  takes me down a rabbit hole of dependencies. I'm looking for suggestions how to solve this problem, please.

Here is my dockerfile:

FROM ghcr.io/ifm/ifm3d:stable
ARG DEBIAN_FRONTEND=noninteractive
ENV TZ=US/EST
RUN sudo apt-get update -y && \
    sudo apt-get install -y apt-utils && \
    sudo apt-get install -y tzdata && \
    sudo apt-get install -y iputils-ping && \
    sudo apt-get install -y net-tools && \
    sudo apt-get install -y iproute2
RUN sudo apt-get install -y --no-install-recommends \
    build-essential \
    git \
    cmake \
    coreutils \
    ###libgflags-dev \
    findutils && \
    sudo apt-get autoclean && \
    sudo apt-get autoremove && \
    sudo apt-get clean && \
    sudo rm -rf /var/lib/apt/lists/*
###RUN sudo apt-get install libgflags-dev

WORKDIR /msclient.d
COPY ms.d ./ms.d
COPY app.d ./app.d
COPY CMakeLists.txt .

WORKDIR build
RUN sudo cmake ..

###ENTRYPOINT ["/bin/bash", "-l"]
###CMD /bin/bash

Here is my CMakeLists.txt:

cmake_minimum_required (VERSION 3.13)

project (MSClient VERSION 0.1.0)

find_package (glog CONFIG REQUIRED PATHS /usr/lib/cmake)
if (NOT glog_FOUND)
   message (FATAL_ERROR "glog was not found. Bye...")
endif()

find_package (ifm3d CONFIG REQUIRED COMPONENTS device framegrabber)
if (NOT ifm3d_FOUND)
   message (FATAL_ERROR "ifm3d was not found. Bye...")
endif()

add_subdirectory(ms.d)

add_subdirectory(app.d)

target_link_libraries (${APP_TARGET} PUBLIC
	ifm3d::device
        ifm3d::framegrabber
	${MSCLIENT_LIB}
        pthread
        )

**Comment by lola-masson** [comment]
Hi @KtGunn,

Are you using glog in your app? It seems like this error might not be related to ifm3d. We typically do not need to install libgflags-dev to run applications using ifm3d. Maybe you can provide the full output of your build command?

**Reply by KtGunn** [reply]
No I am not. My understanding is that the request for glog/gflags comes from 'find_package(ifm3d...)' in my CMakeLists.txt. The ifm3d install in /usr/lib/cmake/ifm3d-1.1.1/ifm3d-config.cmake in the ghcr.io... image looks for glog. Then glog looks for gflags (I think). I might possibly be using the wrong ifm3d image.

**Comment by lola-masson** [comment]
Well the first thing I notice is you are not specifying the architecture you want in the base ifm3d image. Since you are building on linux/amd64, it will default to the docker image for this architecture. You should specify the architecture, for instance use: `FROM ghcr.io/ifm/ifm3d:v1.1.1-ubuntu-arm64`. 

But I don't think this will chance anything to the glog dependency issue. I think these should already be in the base image, but to be sure can you try adding `libgoogle-glog0v5` and `libgoogle-glog-dev` to your apt installed packaged? Then, if you get a build error, please provide the whole build output so I can trace the error.

**Reply by KtGunn** [reply]
Thanks for the pointer about the FROM line, Lola. But as you suspected that did not solve the problem. I added the two 'apt-get install' you suggested and that concluded in a problem I've seen in all my tries: some issue with 'libcurl'. I'm including screen shots of  the build trace instead of copying text as I hope this is easier for you to read. 

![glog-trace-1](https://user-images.githubusercontent.com/6865198/222760203-60e521e7-cf4a-4d2e-a374-54b3b3c56c64.png)


![glog-trace-2](https://user-images.githubusercontent.com/6865198/222760369-92cadf6a-3ce3-43ac-b086-0b963a9775f3.png)

Thanks for your help.

PS I figured I should include the Dockerfile also just in case.


![Dockerfile-glog-trace](https://user-images.githubusercontent.com/6865198/222761626-37d73dae-9e9d-4859-8996-acf4a62e85b3.png)

**Comment by lola-masson** [comment, marked as answer]
I just gave the build a try, and I am able to build a c++ application from this minimal example:

- Dockerfile:
```docker 
FROM ghcr.io/ifm/ifm3d:v1.1.1-ubuntu-arm64

ARG DEBIAN_FRONTEND=noninteractive
RUN sudo apt update && sudo apt install -y\
        build-essential \
        cmake \
        coreutils \
        libgoogle-glog0v5 \
        libgoogle-glog-dev

COPY examples/o3r/ifm3d_playground /home/ifm/ifm3d_playground

WORKDIR /home/ifm/ifm3d_playground/build

RUN cmake .. && cmake --build .
```
- C++ example: I am using [this](https://github.com/ifm/ifm3d/blob/v1.1.1/examples/o3r/ifm3d_playground/ifm3d_playground.cpp) example.
- Build command:
```bash
$ docker build --no-cache --file docker/ifm-robotics/example/Dockerfile .
[+] Building 113.1s (10/10) FINISHED                                                                             
 => [internal] load build definition from Dockerfile                                                        0.0s
 => => transferring dockerfile: 416B                                                                        0.0s
 => [internal] load .dockerignore                                                                           0.0s
 => => transferring context: 107B                                                                           0.0s
 => [internal] load metadata for ghcr.io/ifm/ifm3d:v1.1.1-ubuntu-arm64                                      0.5s
 => CACHED [1/5] FROM ghcr.io/ifm/ifm3d:v1.1.1-ubuntu-arm64@sha256:3cba3f7c1ed492793f8a139100e2f8cf9a0114c  0.0s
 => [internal] load build context                                                                           0.0s
 => => transferring context: 1.58kB                                                                         0.0s
 => [2/5] RUN sudo apt update && sudo apt install -y        build-essential         cmake         coreuti  80.2s
 => [3/5] COPY examples/o3r/ifm3d_playground /home/ifm/ifm3d_playground                                     0.0s
 => [4/5] WORKDIR /home/ifm/ifm3d_playground/build                                                          0.0s
 => [5/5] RUN cmake .. && cmake --build .                                                                  31.1s
 => exporting to image                                                                                      1.3s
 => => exporting layers                                                                                     1.3s
 => => writing image sha256:f856ce39a3c0595cd288352a78d33fb971861dca43e667f339653d7e34e656a7                0.0s
```

Maybe you can start with this, adding components one at a time so we can understand exactly which dependency is causing issues.

**Reply by KtGunn** [reply]
Good news, Lola. I set out to implement your suggestion. Did not do so fully but took it far enough to see it was going to work on my host as it did on yours. Long story short, I did two things. First, I deleted the ifm image from my host entirely and pulled it in fresh. Probably unnecessary. Second, I eliminated the superfluous boilerplate installs from top of my Docker file. I kept the two glog installs you suggested earlier. That built ok. Subsequently I have discovered the 'tzdata' package doesn't play nice. The others do.
Until my next stumbles, thank you for your help.
